sw/device/lib/runtime/irq.c

```diff
@@ -0,0 +1,150 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/runtime/irq.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdnoreturn.h>
+
+#include "sw/device/lib/uart.h"
+
+#define IRQ_ABI __attribute__((aligned(4), interrupt))
+
+size_t irq_type_scratch_area = 0;
+IRQ_ABI void irq_internal_dispatch(void) {
+  irq_handler_t handler =
+      irq_vtable[irq_type_scratch_area / sizeof(irq_handler_t)];
+  uart_send_str("Trampolining to interrupt handler at 0x");
+  uart_send_uint((uintptr_t)handler, 32);
+  uart_send_str(".\n");
+
+  handler();
+  uart_send_str("Interrupt handler done.\n");
+}
+
+static void noreturn irq_software_default(void) {
```

Comment:
I don't like the `irq_software_default` naming, it lacks a verb (okay it has a verb, but our goal is to make sure our software never defaults 😉).  `irq_handle_software_interrupt` is wordy, but does actually communicate what it does.

Reply:
yeah the first time around there was a lively discussion on having a specific naming..i think there were strong feelings about having the terms 'irq_handle' specifically.

Reply:
Well, the point is that these are overrideable defaults, not so much user-accessible functions.

I don't really think of this as a DIF so much as just a runtime support module.